### PR TITLE
[feat] differentiable solarization

### PR DESCRIPTION
**Pull request by paulgavrikov**
#### Changes
Based on #2490 I implemented a `torch.where` approximation based on a sigmoid. This allows `kornia.enhance.solarize` to be differentiable with respect to the parameters (and could be used in other functions with `torch.where`). The relaxation can be triggered by a parameter and is disabled by default. 

(Haven't added test cases)

#### Type of change
- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/enhance/adjust.py [hunk 1]
```diff
@@ -595,15 +595,18 @@ def adjust_log(image: Tensor, gain: float = 1, inv: bool = False, clip_output: b
     return img_adjust
 
 
-def _solarize(input: Tensor, thresholds: Union[float, Tensor] = 0.5) -> Tensor:
+def _solarize(input: Tensor, thresholds: Union[float, Tensor] = 0.5, relaxation: Optional[str] = None) -> Tensor:
```
Comment (edgarriba) [inline kornia/enhance/adjust.py:598]:
```suggestion
def _solarize(input: Tensor, thresholds: Union[float, Tensor] = 0.5, relaxation: str | None = None) -> Tensor:
```
then `from __future__ import annotations`

Comment (paulgavrikov) [reply kornia/enhance/adjust.py]: done

File: kornia/enhance/adjust.py [hunk 2]
```diff
@@ -614,19 +618,25 @@ def _solarize(input: Tensor, thresholds: Union[float, Tensor] = 0.5) -> Tensor:
     if not isinstance(thresholds, (float, Tensor)):
         raise TypeError(f"The factor should be either a float or Tensor. Got {type(thresholds)}")
 
+    if relaxation is not None and relaxation not in ['sigmoid']:
```
Comment (johnnv1) [inline kornia/enhance/adjust.py:621]: You can also use the KORNIA_CHECK API here https://github.com/kornia/kornia/blob/master/kornia/core/check.py#L85
```suggestion
    if relaxation is not None and relaxation not in {'sigmoid'}:
```

File: kornia/enhance/adjust.py [hunk 3]
```diff
@@ -614,19 +618,25 @@ def _solarize(input: Tensor, thresholds: Union[float, Tensor] = 0.5) -> Tensor:
...
             raise AssertionError(f"thresholds must be a 1-d vector of shape ({input.size(0)},). Got {thresholds}")
         # TODO: I am not happy about this line, but no easy to do batch-wise operation
         thresholds = thresholds.to(input.device).to(input.dtype)
         thresholds = torch.stack([x.expand(*input.shape[-3:]) for x in thresholds])
 
-    return torch.where(input < thresholds, input, 1.0 - input)
+    if relaxation is None:
+        img_adjust = torch.where(input < thresholds, input, 1.0 - input)
+    elif relaxation == 'sigmoid':
+        img_adjust = _torch_diff_where(input, thresholds, input, 1.0 - input)
```
Comment (johnnv1) [inline kornia/enhance/adjust.py:634]: can you add test cases for this?

**Comment by johnnv1** [comment]
hey @paulgavrikov, are you still interested/time in finishing this patch?

**Comment by johnnv1** [comment]
as part of https://github.com/kornia/kornia/discussions/2719 plan, we have deleted the `master` in favor of `main`, if anyone has an interest in keep/restart working on this, let me know to help you.
